### ISSUE TITLE
Pin3 pulled low on power up hangs bootloader

### ISSUE THREAD

**Issue by vgtlndr** [micronucleus/micronucleus#135]
Please see comment below. Sorry, I am new to git.

**Comment by vgtlndr** [comment]
Hi all and thanks for all the work done in this project.
I have created a electronic speedo using a digispark reading a hall sensor on pin3. All other pins are occupied with the microstepping stepping driver, a small oled and a voltage divider @ pin4 for reading the voltage. Cant connect the stepper driver to usb pins as it dies when connecting USB. Oled Pins are fixed @ pin0 and pin2.
After updating from 1.11 to 2.3 the circuit failed to startup properly sometimes, i.e. user script is not executed. Figured out that this is related to hall sensor. If the sensors pulls pin3 LOW, i.e. magnet close to it the digispark hangs on startup. It keeps hänging even if the magnet is removed and pin3 can be pulled up be the internal pullup. Moving away the magnet the circuit starts fine after a power cycle.
No issues with the old 1.11 so downgrade fixed it but I would love to have the extra 200 free bytes. Didnt dare to use aggressive though. Entry mode is ENTRY_ALWAYS, config is t85_default.
Any chance to fix this?
Regards, Hendrik

**Comment by AHorneffer** [comment]
Just so that I understand you correctly: you need to use the USB pins as an input, and you disconnect the inputs from the digispark when updating the program? 
Also when you use pin PB4 as an analog input, you should think if the attached resistor and zener diode will mess with your data. (You sound like it doesn't, but it can't hurt to check.)

But your actual problem is that when the digispark is not connected to USB, then it will not start the user program if pin PB3 is pulled low, correct? (If it is connected to USB and PB3 is pulled low, then you have broken USB communication and all bets are off.)

**Comment by ArminJo** [comment]
Hello vgtlndr,
as far as I can see it is the same situation as in issue #123 
I use therefore a modified version from my [fork ](https://github.com/ArminJo/micronucleus/blob/master/firmware/releases/t85_no_pullup.hex). The modification is explained in #123 .
Maybe you can check if your problem vanishes with this version?

**Comment by vgtlndr** [comment]
Hi and thanks for the comments.

I do not disconnect the circuit when hooking up the digispark to pc usb for updating/programming but it is powered off of course. Circuit does not interfere with programming, loading a new script works well. I had to wire it up this way as I had not enough space in the speedo to make the digispark pluggable.
Hall sensor will not pull PB3 to low as it is powered off and lifted away from the magnet. Script starts after programming when still connected to usb but doesnt do anything useful of course.
Next steps: disconnect from usb, put it back into the speedo and power it on to normal operation.
If during power up the hall sensor is pulled low by the magnet the bootloader is stuck. If PB3 is not pulled low the script starts.
Can be simulated with an unconnected digispark easily:
- put a simple script on it to turn the led on
- connect ground to GND and PB3
- connect Vcc
-> wont start, bootloader hangs, script is not executed
- leave PB3 unconnected it will start

As you never know where the magnet stops its a bit of a show stopper but micronucleus 1.11 does not show that behaviour and I managed to strip the code to fit it in the smaller free space as a workaround.

Thanks for the advice with PB4 and the voltage divider but the voltmeter function shows a rough bar graph only to assure that the vehicles alternator is still doing its job and works as expected after manual calibration with reference voltage source.

I do not know if its really the same as #123 but I will rwad trough it and  give it a shot.

Thanks again.


Thanks

**Comment by ArminJo** [comment]
I had exactly the same behavior as you describe, and if you compare the 1.11 and the actual sources you will know why. Therefore I ported the old behavior from the 1.11 into the actual source, but this increases code size by 14 bytes. Because I am not able to decide if this can be tolerated, I introduced a #ifdef and build a new variant.
You should definitely give it a try, I bet it will help you too.

**Comment by vgtlndr** [comment]
Hi, I took the time today to test your fork. Works like a charme!

I think it does not only start correctly now with Pin3 pulled low but it skips the USB delay in this case which is even better. I can not control the pin at startup as it depends on the speedo cable / magnet position but probably somebody else can use this for a faster startup. I already reduced the delay from 6000 to 1500 which is just fine ( I have to kickstart the bike anyway ).

Even if it increases the code by 14 bytes the space usable for the script stays at 6522 with t85_default (got a bit confused by that at first). IMHO this fix is essential and should go in the main release.

Btw. I am a bit confused by the new numbering scheme, MN version 2.04 installs firmware 2.3 ?? Changed that to 2.31 for me to distinguish between versions with or without your fix.

However, thanks a lot again!

**Comment by cpldcpu** [comment]
Ooops. "Someone" forgot to increase the minor revision number. Maybe have to sneak that into main. Thanks for pointing out.

**Comment by vgtlndr** [comment]
@cpldcpu: Could you please explain what kind of danger? Just confusing the bootloader or something serious?  

ENTRY_RESET or Jumper wont work in my setup as all pins incl reset are in use (2 for I2C, 2 controlling a microstepping driver, one reading analog from voltage divider and one reading from hall sensor (yes, I repurposed the reset pin).

I had no problems with 1.11 and I hope V2 will work as good as 1.11 thanks to the fix provided by Armin.

**Comment by AHorneffer** [comment]
@vgtlndr @ArminJo: Well, one danger of re-using the USB lines is that you connect some hardware there that isn't part of the USB. So in the worst case you might fry the USB host (usually your computer). Another danger is that your extra hardware interferes with the USB communication. The other way around, your hardware needs to tolerate that the USB lines change state during USB communication and must not get confused by that. Nothing impossible to solve, but one needs to be careful!

**Comment by vgtlndr** [comment]
Thanks for pointing that out. I have connected a 100k/10k voltage divider to PB4 which connects to ground and the Input is left open during programming which works at least with both hosts I tested it with. The hall sensor is not close to a magnet so it does not interfere either.
But I did kill a couple of stepper driver IC before moving them away from the USB pins. I tried to make the digispark pluggable buth space is very limited and it didnt work out.
Anyway everything fine due to Armins Code. This should go in the main at least as an option to be enabled during compile.
Thanks again to everyone. I close this issue..
